Nama : Tegas Gagas Impian Prahita N

In [1]:
import pandas as pd

# Soal Beginner

## No.1

In [2]:
# Masukan syntax jawaban di sini

data = pd.DataFrame({'no'    : [1, 2, 3, 4, 5],
                     'nama_siswa'  : ['Adi', 'Budi', 'Carli', 'Dadan', 'Endang'],
                     'umur': [15, 16, 14, None, 16],
                     'email'  : ['andi@gmail.com', 'budi@gmail.com', 'carli@gmail.com', 'dadan@gmail.com', 'endang@gmail.com'],
                     'score_matematika'  : [80, 90, 100, 95, 97],
                     'score_ekonomi'  : [90, 95, 100, 86, None],
                     'score_olahraga'  : [97, 95, 96, 100, 94]
                    })

data

no nama_siswa  umur             email  score_matematika  score_ekonomi  \
0   1        Adi  15.0    andi@gmail.com                80           90.0   
1   2       Budi  16.0    budi@gmail.com                90           95.0   
2   3      Carli  14.0   carli@gmail.com               100          100.0   
3   4      Dadan   NaN   dadan@gmail.com                95           86.0   
4   5     Endang  16.0  endang@gmail.com                97            NaN   

   score_olahraga  
0              97  
1              95  
2              96  
3             100  
4              94

## No.2 

In [3]:
# Masukan sytnax jawaban di sini

data_sorted = data.sort_values('score_matematika', ascending=False)

data_lowest_score = data_sorted.tail(1)

print(f"Nama siswa dengan score_matematika terkecil adalah {data_lowest_score.at[0,'nama_siswa']}")

Nama siswa dengan score_matematika terkecil adalah Adi


## No.3

In [4]:
# Masukan syntax jawaban di sini

data_siswa_gmail = data[data.email.astype(str).str.lower().str.contains("gmail") & data.score_ekonomi.notnull()]

umur_mean = data_siswa_gmail.umur.mean()

print(f"Umur rata-rata dari siswa di 'data_siswa_gmail' adalah {umur_mean}")

Umur rata-rata dari siswa di 'data_siswa_gmail' adalah 15.0


## No.4

In [5]:
# Masukan syntax jawaban di sini

additional_data = pd.DataFrame({'no'    : [6, 7],
                     'nama_siswa'  : ['Fauzan', 'Guntur'],
                     'umur': [15, 16],
                     'email'  : ['fauzan@gmail.com', 'guntur@gmail.com'],
                     'score_matematika'  : [80, 90],
                     'score_ekonomi'  : [90, 95],
                     'score_olahraga'  : [97, 95]
                    })

# new_data = data.append(additional_data, ignore_index=True)
new_data = pd.concat([data, additional_data], ignore_index=True)
# print(new_data)

umur_med = new_data.umur.median()

print(f"Median umur mereka sekarang adalah {umur_med}")

Median umur mereka sekarang adalah 15.5


## No.5

In [6]:
##### Masukan syntax jawaban di sini

data_address = pd.DataFrame({'nama'    : ['Adi', 'Budi', 'Dadan', 'Endang', 'Guntur', 'Halim'],
                     'alamat'  : ['DKI Jakarta', 'DKI Jakarta', 'Banten', 'Banten', 'Jawa Barat', 'Aceh']
                    })


# merge data frame “new_data”
data_merge = new_data.merge(data_address, left_on='nama_siswa', right_on='nama', how='outer')

# umur rata-rata siswa per alamat mereka berasal
umur_add_mean = data_merge.groupby('alamat').agg({'umur' : 'mean'})

# x = umur_add_mean.reset_index()

# y = umur_add_mean.umur.min()

# z = x.loc[x['umur'] == y]['alamat'].values.tolist()

print(f"Umur rata-rata mereka yang paling kecil pada alamat  {umur_add_mean['umur'].idxmin()}")

Umur rata-rata mereka yang paling kecil pada alamat  DKI Jakarta


# Soal Intermediate

## No.6

In [7]:
# Masukan syntax jawaban di sini

airlines = pd.read_excel(r'C:\Users\tegas\Documents\rakamin\Homework\Vemby\data_airlines.xlsx')
airports = pd.read_excel(r'C:\Users\tegas\Documents\rakamin\Homework\Vemby\data_airports.xlsx')
flights = pd.read_excel(r'C:\Users\tegas\Documents\rakamin\Homework\Vemby\data_flights.xlsx')

In [ ]:
flights.head()
# flights.FLIGHT_NUMBER.value_counts()

In [8]:
df = flights.groupby('MONTH').agg({'FLIGHT_NUMBER' : 'count',
                                    'CANCELLED' : 'sum',
                                    'ELAPSED_TIME' : ['mean', 'max', 'min']}).reset_index()

df.columns = ['bulan','total_penerbangan','total_pembatalan','waktu_tempuh_rerata','waktu_tempuh_max','waktu_tempuh_min']

df

bulan  total_penerbangan  total_pembatalan  waktu_tempuh_rerata  \
0       1                798                20           132.153548   
1       2                715                30           140.375731   
2       3                877                22           142.357562   
3       4                824                 9           138.420859   
4       5                840                 5           136.516827   
5       6                839                18           139.057317   
6       7                937                14           141.110990   
7       8                887                 8           142.948689   
8       9                787                 1           137.667516   
9      10                856                 2           134.366940   
10     11                822                 8           139.198032   
11     12                818                17           144.340426   

    waktu_tempuh_max  waktu_tempuh_min  
0              428.0              31.0  
1              408.0              24.0  
2              622.0              30.0  
3              570.0              31.0  
4              466.0              34.0  
5              398.0              32.0  
6              511.0              30.0  
7              655.0              27.0  
8              495.0              34.0  
9              388.0              31.0  
10             392.0              32.0  
11             453.0              33.0

**Insight**

Selama tahun 2015, pada bulan Juli memiliki jadwal penerbangan tersibuk yaitu 937 dan bulan Februari memiliki jadwal penerbangan paling sedikit yaitu 715. Ketika bulan Februari juga terdapat pembatalan penerbangan yang paling banyak dibandingkan bulan lainnya. Bulan September hampir seluruh penerbangan diberangkatkan dan hanya terdapat 1 pembatalan saja.

Penerbangan pada bulan Juli merupakan yang teramai dikarenakan puncak libur musim panas sehingga customer banyak melakukan berpergian. Pada bulan Januari - Maret memiliki jumlah pembatalan penerbangan yang relatif lebih besar dibandingkan bulan lainnya (diatas 20 kali). Hal ini dimungkinkan karena adanya cuaca buruk saat penerbangan di musim dingin.

Waktu tempuh penerbangan (elapsed time) secara umum konstan untuk setiap bulan berkisar 140 menit, dengan waktu terpendek berkisar 30 menit dan waktu terpanjang berkisar 380 hingga 650 menit. Waktu tempuh yang bervariasi bergantung terhadap beberapa kodisi seperti cuaca dan tingkat kepadatan bandara.

## No.7

In [ ]:
# Masukan syntax jawaban di sini
airports[airports.IATA_CODE.astype(str).str.contains("JFK")].head()

In [ ]:
airports[airports.IATA_CODE.astype(str).str.contains("ECP")].head()

In [9]:
airports.AIRPORT.isnull().any()

False

In [10]:
tmp = flights.merge(airports, left_on='ORIGIN_AIRPORT',right_on='IATA_CODE', how='left')
df = tmp.merge(airports, left_on='DESTINATION_AIRPORT',right_on='IATA_CODE', how='left')

df = df.rename(columns = {'AIRPORT_x' : 'ORI_AIRPORT', 'AIRPORT_y' : 'DEST_AIRPORT'})

df[['AIRLINE', 'FLIGHT_NUMBER', 'ORI_AIRPORT', 'DEST_AIRPORT']]

AIRLINE  FLIGHT_NUMBER  \
0         UA           1475   
1         AA            227   
2         AA            357   
3         MQ           3459   
4         OO           6559   
...      ...            ...   
9995      AS            579   
9996      DL           2172   
9997      NK            801   
9998      EV           4093   
9999      WN           5119   

                                            ORI_AIRPORT  \
0                  Chicago O'Hare International Airport   
1               Dallas/Fort Worth International Airport   
2            Baltimore-Washington International Airport   
3                           Miami International Airport   
4              Phoenix Sky Harbor International Airport   
...                                                 ...   
9995  San Diego International Airport (Lindbergh Field)   
9996                        Tampa International Airport   
9997                       Detroit Metropolitan Airport   
9998         Savannah/Hilton Head International Airport   
9999                     Portland International Airport   

                                           DEST_AIRPORT  
0                         Orlando International Airport  
1          Minneapolis-Saint Paul International Airport  
2                           Miami International Airport  
3     Pensacola International Airport (Pensacola Gul...  
4                    Palm Springs International Airport  
...                                                 ...  
9995                     Portland International Airport  
9996   Hartsfield-Jackson Atlanta International Airport  
9997                      Orlando International Airport  
9998               Chicago O'Hare International Airport  
9999         John Wayne Airport (Orange County Airport)  

[10000 rows x 4 columns]

In [11]:
df.duplicated().any()

False

In [12]:
df_n = df[df.ORI_AIRPORT.isin(['John F. Kennedy International Airport (New York International Airport)',
                               'Northwest Florida Beaches International Airport'])]

df_n[['AIRLINE', 'FLIGHT_NUMBER', 'ORI_AIRPORT', 'DEST_AIRPORT']]

AIRLINE  FLIGHT_NUMBER  \
44        AA           1275   
50        B6            353   
86        UA            445   
163       AA              1   
235       MQ           3378   
...      ...            ...   
9808      AA             19   
9896      DL            435   
9902      B6           1983   
9903      B6            518   
9952      AA           2445   

                                            ORI_AIRPORT  \
44    John F. Kennedy International Airport (New Yor...   
50    John F. Kennedy International Airport (New Yor...   
86    John F. Kennedy International Airport (New Yor...   
163   John F. Kennedy International Airport (New Yor...   
235   John F. Kennedy International Airport (New Yor...   
...                                                 ...   
9808  John F. Kennedy International Airport (New Yor...   
9896  John F. Kennedy International Airport (New Yor...   
9902  John F. Kennedy International Airport (New Yor...   
9903  John F. Kennedy International Airport (New Yor...   
9952  John F. Kennedy International Airport (New Yor...   

                                          DEST_AIRPORT  
44                               Cyril E. King Airport  
50                    Palm Beach International Airport  
86                   Los Angeles International Airport  
163                  Los Angeles International Airport  
235          Ronald Reagan Washington National Airport  
...                                                ...  
9808                 Los Angeles International Airport  
9896               San Francisco International Airport  
9902                     Orlando International Airport  
9903  Gen. Edward Lawrence Logan International Airport  
9952                     Orlando International Airport  

[181 rows x 4 columns]

In [13]:
df_cl = df_n.groupby(['ORI_AIRPORT','DEST_AIRPORT']).agg({'FLIGHT_NUMBER' : 'count','DISTANCE' : 'mean'}).reset_index()

df_cl.columns = ['asal','tujuan','total_penerbangan','jarak_rerata']

df_cl.tail(10)

asal  \
36  John F. Kennedy International Airport (New Yor...   
37  John F. Kennedy International Airport (New Yor...   
38  John F. Kennedy International Airport (New Yor...   
39  John F. Kennedy International Airport (New Yor...   
40  John F. Kennedy International Airport (New Yor...   
41    Northwest Florida Beaches International Airport   
42    Northwest Florida Beaches International Airport   
43    Northwest Florida Beaches International Airport   
44    Northwest Florida Beaches International Airport   
45    Northwest Florida Beaches International Airport   

                                              tujuan  total_penerbangan  \
36               San Francisco International Airport                 23   
37              Seattle-Tacoma International Airport                  3   
38           Southwest Florida International Airport                  3   
39                       Tampa International Airport                  7   
40           Washington Dulles International Airport                  1   
41                                 Dallas Love Field                  2   
42              George Bush Intercontinental Airport                  1   
43  Hartsfield-Jackson Atlanta International Airport                  3   
44                   Nashville International Airport                  1   
45                          William P. Hobby Airport                  1   

    jarak_rerata  
36        2586.0  
37        2422.0  
38        1074.0  
39        1005.0  
40         228.0  
41         674.0  
42         572.0  
43         240.0  
44         401.0  
45         571.0

In [14]:
tmp = pd.pivot_table(df_cl, 
                     index = ['tujuan'],
                     columns = ['asal'],
                     values = ['total_penerbangan','jarak_rerata']
                    )

tmp.columns = ['jarak_rerata_jfk', 'jarak_rerata_ecp', 'total_penerbangan_jfk', 'total_penerbangan_ecp']

In [15]:
idx_jfk_max = tmp.total_penerbangan_jfk.idxmax()
idx_ecp_max = tmp.total_penerbangan_ecp.idxmax()

print(f"Destinasi airport mana yang paling banyak jumlah penerbangannya dari JFK yaitu: {idx_jfk_max}")
print(f"Destinasi airport mana yang paling banyak jumlah penerbangannya dari ECP yaitu: {idx_ecp_max}")

Destinasi airport mana yang paling banyak jumlah penerbangannya dari JFK yaitu: San Francisco International Airport
Destinasi airport mana yang paling banyak jumlah penerbangannya dari ECP yaitu: Hartsfield-Jackson Atlanta International Airport


Destinasi airport mana yang paling banyak jumlah penerbangannya yaitu:

| Asal | Tujuan | Jumlah | Jarak tempuh |
|---|---|---|---|
| John F. Kennedy | San Francisco | 23 | 2586.0 |
| Northwest Florida Beaches | Hartsfield-Jackson Atlanta | 3 | 240.0 |

# Soal Advanced

## No.8

In [16]:
# Masukan syntax jawaban di sini

df_a = pd.read_csv(r'C:\Users\tegas\Documents\rakamin\Homework\Vemby\01 telecom_revenue.csv')
df_b = pd.read_csv(r'C:\Users\tegas\Documents\rakamin\Homework\Vemby\02 telecom_usage.csv')

df_1 = df_a[(df_a.Occupation.isin(['Professional', 'Student', 'Crafts'])) & (df_a.MonthlyRevenue > 10)] 
df_2 = df_b[(df_b.UnansweredCalls > df_b['BlockedCalls']) & 
             ((df_b.CustomerCareCalls.astype(str).str.contains("^8")) | 
              (df_b.CustomerCareCalls.astype(str).str.contains("^4")) |
              (df_b.CustomerCareCalls.astype(str).str.contains("^2")))]  

In [ ]:
df_1

In [ ]:
df_2

In [17]:
df_m = df_1.merge(df_2, how='inner', on='CustomerID').sort_values('BlockedCalls', ascending=False)

In [18]:
df_n = df_m.groupby(['Occupation']).agg({'CustomerID' : 'nunique', 'DroppedCalls' : 'mean'})

df_n.duplicated().any()

False

In [19]:
print(f" Rata-rata DroppedCalls paling besar terdapat pada pekerjaan {df_n.DroppedCalls.idxmax()}")

 Rata-rata DroppedCalls paling besar terdapat pada pekerjaan Crafts
